src/cli/index.js

```diff
@@ -443,7 +443,8 @@ export function main({
     return errorReportLoc;
   }
 
-  const cwd = findProjectRoot(commander.cwd);
+  const shouldUseProjectRoot = commandName !== 'init';
```

Comment:
This is a good start! That said I think at this point we already know the command itself and this information is better stored on the command class so here's what I'd do:

 - Put a `shouldRunInCurrentCwd` property on the `init` command here: https://github.com/yarnpkg/yarn/blob/master/src/cli/commands/init.js#L25 (you'd need to split the `export` and `function` definitions there to be able to do this.
 - on the line below do this instead:
   ```
    const cwd = command.shouldRunInCurrentCwd ? commander.cwd : findProjectRoot(commander.cwd);
    ```

You can add tests to https://github.com/yarnpkg/yarn/blob/master/__tests__/commands/init.js and the fixtures for these commands are at https://github.com/yarnpkg/yarn/tree/master/__tests__/fixtures/init

I'd add a new fixture with two nested folders where the higher one has a package.json and then run the command in the nested one which doesn't have `package.json`.